Add support for WebAssembly GC recursion groups

takikawa opened the pull request:
#### 5f3859595e5e1e337beac77a6b45ec05421e1842
<pre>
Add support for WebAssembly GC recursion groups
<a href="https://bugs.webkit.org/show_bug.cgi?id=239666">https://bugs.webkit.org/show_bug.cgi?id=239666</a>

Reviewed by Justin Michaud.

Add support for the Wasm GC proposal&apos;s recursion groups. These are
a new kind of TypeDefinition that allows for recursive type
definitions. Recursive type index references are only allowed with
recursion group definitions.

Recursion groups can be stored hash-consed like other type
definitions, but to successfully compare them they must store several
definitions in the type store:

  - The recursion group itself (rec (type ...) ...)
  - Each type component in the group (type ...), where any potential
    recursive references are represented as a special value type.
  - Projections into the recursion group ((rec ...).&lt;i&gt;)

Projections can be expanded into the underlying function/struct/etc
type when the validator needs to examine the structure of the type.
This operation exposes the type component in the group, and replaces
recursive references with the appropriate projection (which can then
be expanded if required, and so on).

If expansion becomes a bottleneck, it is easy to memoize it with the
addition of a TypeIndex to TypeIndex mapping in the type store.

Generally, the expansion operation is needed when code relies on the
structure of the underlying type. For functions, these uses usually
look like `signature.as&lt;FunctionSignature&gt;()`. The signature *must* be
expanded before such a conversion is called.

When signatures need to be compared for equality, such as for
call_indirect or for matching function import/export, the comparison
should generally be done by the type index of the projection. That is,
*not* by equality of the underlying/expanded type.

Recursive references are not represented in the binary format and are
internal to the semantics and implementation. To avoid using extra
opcodes, they are internally represented as a type that uses the
opcode for recursion group (TypeKind::Rec), which cannot normally be
used for value types in the binary format, with a heap type that
encodes the index into the recursion group&apos;s type list.

* JSTests/wasm/gc/rec.js: Added.
(module):
(testRecDeclaration):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::AirIRGenerator):
(JSC::Wasm::AirIRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isValueType):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.cpp:
(JSC::Wasm::FunctionCodeBlockGenerator::addSignature):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::splitStack):
(JSC::Wasm::FunctionParser&lt;Context&gt;::FunctionParser):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parse):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::initElementSegment):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
(JSC::Wasm::LLIntGenerator::addCallIndirect):
(JSC::Wasm::LLIntGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmLimits.h:
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::Parser&lt;SuccessType&gt;::Parser):
(JSC::Wasm::Parser&lt;SuccessType&gt;::parseHeapType):
(JSC::Wasm::Parser&lt;SuccessType&gt;::parseValueType):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseElement):
(JSC::Wasm::SectionParser::parseRecursionGroup):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCallIndirect):
(JSC::LLInt::doWasmCallRef):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeDefinition::dump const):
(JSC::Wasm::RecursionGroup::toString const):
(JSC::Wasm::RecursionGroup::dump const):
(JSC::Wasm::Projection::toString const):
(JSC::Wasm::Projection::dump const):
(JSC::Wasm::computeRecursionGroupHash):
(JSC::Wasm::computeProjectionHash):
(JSC::Wasm::TypeDefinition::hash const):
(JSC::Wasm::TypeDefinition::tryCreateFunctionSignature):
(JSC::Wasm::TypeDefinition::tryCreateStructType):
(JSC::Wasm::TypeDefinition::tryCreateRecursionGroup):
(JSC::Wasm::TypeDefinition::tryCreateProjection):
(JSC::Wasm::TypeDefinition::substitute):
(JSC::Wasm::TypeDefinition::replacePlaceholders const):
(JSC::Wasm::TypeDefinition::expand const):
(JSC::Wasm::FunctionParameterTypes::equal):
(JSC::Wasm::StructParameterTypes::equal):
(JSC::Wasm::RecursionGroupParameterTypes::hash):
(JSC::Wasm::RecursionGroupParameterTypes::equal):
(JSC::Wasm::RecursionGroupParameterTypes::translate):
(JSC::Wasm::ProjectionParameterTypes::hash):
(JSC::Wasm::ProjectionParameterTypes::equal):
(JSC::Wasm::ProjectionParameterTypes::translate):
(JSC::Wasm::TypeInformation::typeDefinitionForRecursionGroup):
(JSC::Wasm::TypeInformation::typeDefinitionForProjection):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::RecursionGroup::RecursionGroup):
(JSC::Wasm::RecursionGroup::typeCount const):
(JSC::Wasm::RecursionGroup::type const):
(JSC::Wasm::RecursionGroup::getType):
(JSC::Wasm::RecursionGroup::storage):
(JSC::Wasm::RecursionGroup::storage const):
(JSC::Wasm::Projection::Projection):
(JSC::Wasm::Projection::recursionGroup const):
(JSC::Wasm::Projection::index const):
(JSC::Wasm::Projection::getRecursionGroup):
(JSC::Wasm::Projection::getIndex):
(JSC::Wasm::Projection::storage):
(JSC::Wasm::Projection::storage const):
(JSC::Wasm::TypeDefinition::TypeDefinition):
(JSC::Wasm::TypeDefinition::allocatedRecursionGroupSize):
(JSC::Wasm::TypeDefinition::allocatedProjectionSize):
* Source/JavaScriptCore/wasm/WasmTypeDefinitionInlines.h:
(JSC::Wasm::TypeInformation::getFunctionSignature):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: <a href="https://commits.webkit.org/253491@main">https://commits.webkit.org/253491@main</a>
</pre>

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -123,27 +125,93 @@ class StructType {
+
+    TypeIndex& getType(RecursionGroupCount i) { ASSERT(i < typeCount());; return *storage(i); }
+    TypeIndex* storage(RecursionGroupCount i) { return i + m_payload; }
+    TypeIndex* storage(RecursionGroupCount i) const { return const_cast<RecursionGroup*>(this)->storage(i); }
```
Shouldn't this be `const TypeIndex* storage(RecursionGroupCount i) const`?

takikawa replied:
Thanks for catching, fixed.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -123,27 +125,93 @@ class StructType {
+    TypeIndex& getRecursionGroup() { return *storage(0); }
+    ProjectionIndex& getIndex() { return *reinterpret_cast<ProjectionIndex*>(storage(1)); }
+    TypeIndex* storage(uint32_t i) { ASSERT(i <= 1); return i + m_payload; }
+    TypeIndex* storage(uint32_t i) const { return const_cast<Projection*>(this)->storage(i); }
```
Ditto on const return type

kmiller68 commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
```diff
@@ -123,27 +125,93 @@ class StructType {
+    RecursionGroupCount m_typeCount;
+};
+
+class Projection {
```
It might be worth having a comment describing what a Projection is for people who are not familiar with academic type systems.

takikawa replied:
Done.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -724,6 +735,51 @@ auto SectionParser::parseStructType(uint32_t position, RefPtr<TypeDefinition>& s
+            break;
+        }
+        default:
+            return fail(i, "th Type is non-Func ", typeKind);
```
I think this should be non-Func, non-Struct in the error message?

takikawa replied:
Done.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -724,6 +735,51 @@ auto SectionParser::parseStructType(uint32_t position, RefPtr<TypeDefinition>& s
+        case TypeKind::Struct: {
+            WASM_FAIL_IF_HELPER_FAILS(parseStructType(i, signature));
+            break;
+        }
```
Can we add a FIXME to add Array types?

takikawa replied:
Done.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -724,6 +735,51 @@ auto SectionParser::parseStructType(uint32_t position, RefPtr<TypeDefinition>& s
+    Vector<TypeIndex> types;
+    WASM_PARSER_FAIL_IF(!types.tryReserveCapacity(typeCount), "can't allocate enough memory for recursion group ", typeCount, " entries");
+
+    RecursionGroupGuard guard = RecursionGroupGuard(this, RecursionGroupInformation { true, m_info->typeCount(), m_info->typeCount() + typeCount });
```
I think this should just be `SetForScope recursionGroupInfo(m_recursionGroupInformation, { true, m_info->typeCount(), m_info->typeCount() + typeCount })`

takikawa replied:
I did this with some additional type info and adjusted `m_recursionGroupInformation` to be `protected`.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmParser.h:
```diff
@@ -64,6 +64,30 @@ class Parser {
+
+    private:
+        Parser* m_parser;
+    };
```
I don't think you need this, see below.

takikawa replied:
Removed now.

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -265,7 +265,7 @@ class LLIntCallee final : public Callee {
     LLIntTierUpCounter& tierUpCounter() { return m_tierUpCounter; }
 
-    const FunctionSignature& signature(unsigned index) const
+    const TypeDefinition& signature(unsigned index) const
```
I think this is wrong. I believe all wasm callees are still FunctionSignatures.

takikawa replied:
AFAICT this is required due to the semantics of recursion groups. The reason is that with the addition of recursion groups, in general all type signatures have to be compared by projection, even if you know it will be a function signature underneath.

The reason is you may have function types like the following:

```
(rec (type (func $f1)) (type (struct ...)))

(rec (type (func $f2)))
```

Even though `$f1` and `$f2` look the same, and are indeed the same if you unfold the projections, they are not equal types because they are in different recursion groups.

(it seems a bit silly in this very simple case, but you can construct more interesting examples where `$f1` and `$f2` would unfold to the same type that, say, reference another member of the recursion group. And where these other members are not equal types.)

justinmichaud replied:
@kmiller68 was trying to explain this to me, so I am wondering if I can ask few clarifying questions: 

```
(rec $r1
    (type $tree (struct (field i32) (field (ref $forest))))
    (type $forest (struct (field (ref $tree)) (field (ref $forest)))))
 
(type $s1 (struct (field (ref $tree))))
```

1) Types are stored as instances of TypeInformation, and expanded to produce a FunctionSignature (including for struct types)?

1) Projection in this case means taking a TypeDefinition, and selecting an element of it without any substitution? So:

$r1 is given a type index
$s1 is represented by TypeDefinition(StructType(Projection($r1, 0)))
$tree is TypeDefinition(Projection($r1, 0))

1) Expanding replaces a projection with a new projection that has had one step of unrolling applied, with the recursion group as the target of the unrolling

1) Unrolling is applying one step of substitution, replacing references to the recursion group with its expansion. So:

expand($tree) <=>
unroll(Projection($r1, 0) with respect to $r1) <=>
TypeDefinition(StructType(i32), TypeDefinition(Projection($r1, 0)))

This new expansion is given its own type index, and comparisons are made by type index?

We need to compare types by type index because the type definitions are iso-recursive? So a TypeDefinition and its .expand() are not the same type? 

If they were equirecursive, then we would have to perform a dfs expanding until we could prove that the types did/did not match?

Sorry to bug you with all these questions, I will continue playing with this patch to see how many of these questions I can answer by myself.

kmiller68 reviewed (commented):
Still reviewing but here are some initial comments

justinmichaud commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -63,12 +64,23 @@ auto SectionParser::parseType() -> PartialResult
 
         WASM_PARSER_FAIL_IF(!signature, "can't allocate enough memory for Type section's ", i, "th signature");
-        m_info->typeSignatures.uncheckedAppend(signature.releaseNonNull());
+        // Recursion group parsing will append the entries itself, as there may
```
Hmm, I wish these cases both acted the same way. 

Consider
```
(struct $a i32)
(rec (struct $b i32))
(struct $c i32)
```
Is it the case that $a != $b != $c, since they are all defined in separate recursion groups? In the parlance of the spec draft, they have different tied context types?

justinmichaud replied:
Either way, it would be nice for all the cases to act the same way.

takikawa replied:
These should actually all be equivalent, I think it's just an oversight/bug in this patch. You can verify this by trying this example in the reference interpreter:

```
(module 
  (type $a (struct (field i32)))
  (rec (type $b (struct (field i32))))
  (type $c (struct (field i32)))

  (func (result (ref null $a)) (ref.null $b))
  (func (result (ref null $a)) (ref.null $c))
  (func (result (ref null $b)) (ref.null $a))
  (func (result (ref null $b)) (ref.null $c))
  (func (result (ref null $c)) (ref.null $a))
  (func (result (ref null $c)) (ref.null $b)))
```

which should validate AFAICT.

The issue in the code is that either we have to normalize structs without a recursion group to be represented with a singleton recursion group, or we have to normalize non-recursive, single-member recursion groups into just a struct definition.

I think the latter approach is probably better, as it avoids the need to unroll projections when they're not needed. I did something similar for normalizing subtyping declarations (the spec says types without an explicit `sub` declaration are treated as having a `sub` with empty parents, just like how types without an explicit `rec` are treated as being wrapped with a `rec`).

takikawa replied:
I just fixed this case in the latest push, and added a test case for the above example. It normalizes non-recursive, single recursion groups as being just the underlying type definition.

Parsing still has multiple code paths for this though, because the binary shorthand may or may not be used.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
```diff
@@ -491,7 +503,7 @@ auto SectionParser::parseElement() -> PartialResult
             Type refType;
             WASM_PARSER_FAIL_IF(!parseRefType(m_info, refType), "can't parse reftype in elem section");
-            WASM_PARSER_FAIL_IF(!refType.isFuncref(), "reftype in element section should be funcref");
+            WASM_PARSER_FAIL_IF(!isFuncref(refType) && refType.isNullable(), "reftype in element section should be funcref");
```
I am confused about the logic here, but also the message is wrong?

takikawa replied:
Yeah, this is unfortunately a bit confusing.

When the typed function references proposal is fully fleshed out, it will support both tables and element sections with concrete function types (as described in https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#summary). But right now, it's restricted to `funcref`.

The confusing thing is that the representation of `funcref` changes a bit between Wasm with reference types and Wasm with typed function references. In the former, it was always a type with `TypeKind::Funcref`. In the latter, it is a `TypeKind::Ref` with a heap type of `func` and the other form becomes a shorthand for this.

The old code `!refType.isFuncref()` is only checking for the shorthand form. This causes some of the tests in this patch to fail, because some of the binary modules use the new form (there is otherwise no connections to recursion groups). The new code is doing the equivalent check but allowing for the non-shorthand `Ref` type form. So there is no change in functionality here; the code is just accounting for a change in representation caused by typed function references.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
```diff
@@ -506,7 +506,7 @@ inline SlowPathReturnType doWasmCallIndirect(CallFrame* callFrame, Wasm::Instanc
 
     const auto& callSignature = CALLEE()->signature(typeIndex);
-    if (callSignature != Wasm::TypeInformation::getFunctionSignature(function.typeIndex))
+    if (callSignature.index() != function.typeIndex)
```
The original check should never find that the modules aren't the same, right?

takikawa replied:
Did you mean the function signatures aren't the same? In that case, no, the check could fail because a Wasm table could be initialized with type `funcref` and the table contents matching the call's function type wouldn't be enforced statically.

(but you're right that it cannot fail if typed function references are used and the table is declared with a concrete function type, as in the examples at https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#examples)

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -117,16 +153,42 @@ static unsigned computeStructTypeHash(size_t fieldCount, const StructField* fiel
 
+static unsigned computeRecursionGroupHash(size_t typeCount, const TypeIndex* types)
+{
+    unsigned accumulator = 0x9cfb89bb;
```
How do we pick these?

takikawa replied:
I've been picking by generating a random hex number, but if there is a more principled way to do it I would be happy to change this.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -147,10 +209,76 @@ RefPtr<TypeDefinition> TypeDefinition::tryCreateStructType(StructFieldCount fiel
     return adoptRef(signature);
 }
 
+RefPtr<TypeDefinition> TypeDefinition::tryCreateProjection()
```
I would love some comments here, and maybe a link to the draft spec where these operations are defined?

takikawa replied:
I added more comments for the various operations here, with some links to the spec.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -147,10 +209,76 @@ RefPtr<TypeDefinition> TypeDefinition::tryCreateStructType(StructFieldCount fiel
+
+const TypeDefinition& TypeDefinition::unroll(TypeIndex projectee) const
+{
+    if (is<FunctionSignature>()) {
```
Why don't we unroll struct definitions?

takikawa replied:
We do have to unroll struct definitions, but currently there is no code that would ever call `unroll` for a struct because no struct operations have been implemented yet. So an unroll on structs would not be exercised in any tests.

How about I add a case for `StructType` here with `ASSERT_NOT_REACHED` and a FIXME comment?

takikawa replied:
I added the assertion and fixme comment.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -260,6 +388,78 @@ struct StructParameterTypes {
+    }
+
+    static void translate(TypeHash& entry, const RecursionGroupParameterTypes& params, unsigned)
+    {
```
What does this do?

takikawa replied:
This is the same as in the `ArrayType` case so I'll link to that comment for explanation: https://github.com/WebKit/WebKit/pull/1128#discussion_r929504379

justinmichaud replied:
Can we get a comment somewhere about this?

takikawa replied:
Added a comment explaining this (and all the other `translate` methods) to the `FunctionSignature` case.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
```diff
@@ -147,10 +209,100 @@ RefPtr<TypeDefinition> TypeDefinition::tryCreateStructType(StructFieldCount fiel
+    return type;
+}
+
+//
```
Can we get a comment here?

takikawa replied:
Added a comment here, and also renamed `unroll` to `replacePlaceholders` because this function didn't actually align very closely with what `unroll` is doing in the spec. The subtyping patch may end up adding an `unroll` again but for now it's not needed.

justinmichaud reviewed (approved):
r=me modulo comment nits

takikawa committed:
Add support for WebAssembly GC recursion groups
https://bugs.webkit.org/show_bug.cgi?id=239666

Reviewed by Justin Michaud.

Add support for the Wasm GC proposal's recursion groups. These are
a new kind of TypeDefinition that allows for recursive type
definitions. Recursive type index references are only allowed with
recursion group definitions.

Recursion groups can be stored hash-consed like other type
definitions, but to successfully compare them they must store several
definitions in the type store:

  - The recursion group itself (rec (type ...) ...)
  - Each type component in the group (type ...), where any potential
    recursive references are represented as a special value type.
  - Projections into the recursion group ((rec ...).<i>)

Projections can be expanded into the underlying function/struct/etc
type when the validator needs to examine the structure of the type.
This operation exposes the type component in the group, and replaces
recursive references with the appropriate projection (which can then
be expanded if required, and so on).

If expansion becomes a bottleneck, it is easy to memoize it with the
addition of a TypeIndex to TypeIndex mapping in the type store.

Generally, the expansion operation is needed when code relies on the
structure of the underlying type. For functions, these uses usually
look like `signature.as<FunctionSignature>()`. The signature *must* be
expanded before such a conversion is called.

When signatures need to be compared for equality, such as for
call_indirect or for matching function import/export, the comparison
should generally be done by the type index of the projection. That is,
*not* by equality of the underlying/expanded type.

Recursive references are not represented in the binary format and are
internal to the semantics and implementation. To avoid using extra
opcodes, they are internally represented as a type that uses the
opcode for recursion group (TypeKind::Rec), which cannot normally be
used for value types in the binary format, with a heap type that
encodes the index into the recursion group's type list.

* JSTests/wasm/gc/rec.js: Added.
(module):
(testRecDeclaration):
* JSTests/wasm/wasm.json:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::AirIRGenerator):
(JSC::Wasm::AirIRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::isValueType):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.cpp:
(JSC::Wasm::FunctionCodeBlockGenerator::addSignature):
* Source/JavaScriptCore/wasm/WasmFunctionCodeBlockGenerator.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::splitStack):
(JSC::Wasm::FunctionParser<Context>::FunctionParser):
(JSC::Wasm::FunctionParser<Context>::parse):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmInstance.cpp:
(JSC::Wasm::Instance::initElementSegment):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::callInformationForCaller):
(JSC::Wasm::LLIntGenerator::callInformationForCallee):
(JSC::Wasm::LLIntGenerator::addArguments):
(JSC::Wasm::LLIntGenerator::addCallIndirect):
(JSC::Wasm::LLIntGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmLimits.h:
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::Parser<SuccessType>::Parser):
(JSC::Wasm::Parser<SuccessType>::parseHeapType):
(JSC::Wasm::Parser<SuccessType>::parseValueType):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseType):
(JSC::Wasm::SectionParser::parseElement):
(JSC::Wasm::SectionParser::parseRecursionGroup):
* Source/JavaScriptCore/wasm/WasmSectionParser.h:
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCallIndirect):
(JSC::LLInt::doWasmCallRef):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeDefinition::dump const):
(JSC::Wasm::RecursionGroup::toString const):
(JSC::Wasm::RecursionGroup::dump const):
(JSC::Wasm::Projection::toString const):
(JSC::Wasm::Projection::dump const):
(JSC::Wasm::computeRecursionGroupHash):
(JSC::Wasm::computeProjectionHash):
(JSC::Wasm::TypeDefinition::hash const):
(JSC::Wasm::TypeDefinition::tryCreateFunctionSignature):
(JSC::Wasm::TypeDefinition::tryCreateStructType):
(JSC::Wasm::TypeDefinition::tryCreateRecursionGroup):
(JSC::Wasm::TypeDefinition::tryCreateProjection):
(JSC::Wasm::TypeDefinition::substitute):
(JSC::Wasm::TypeDefinition::replacePlaceholders const):
(JSC::Wasm::TypeDefinition::expand const):
(JSC::Wasm::FunctionParameterTypes::equal):
(JSC::Wasm::StructParameterTypes::equal):
(JSC::Wasm::RecursionGroupParameterTypes::hash):
(JSC::Wasm::RecursionGroupParameterTypes::equal):
(JSC::Wasm::RecursionGroupParameterTypes::translate):
(JSC::Wasm::ProjectionParameterTypes::hash):
(JSC::Wasm::ProjectionParameterTypes::equal):
(JSC::Wasm::ProjectionParameterTypes::translate):
(JSC::Wasm::TypeInformation::typeDefinitionForRecursionGroup):
(JSC::Wasm::TypeInformation::typeDefinitionForProjection):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::RecursionGroup::RecursionGroup):
(JSC::Wasm::RecursionGroup::typeCount const):
(JSC::Wasm::RecursionGroup::type const):
(JSC::Wasm::RecursionGroup::getType):
(JSC::Wasm::RecursionGroup::storage):
(JSC::Wasm::RecursionGroup::storage const):
(JSC::Wasm::Projection::Projection):
(JSC::Wasm::Projection::recursionGroup const):
(JSC::Wasm::Projection::index const):
(JSC::Wasm::Projection::getRecursionGroup):
(JSC::Wasm::Projection::getIndex):
(JSC::Wasm::Projection::storage):
(JSC::Wasm::Projection::storage const):
(JSC::Wasm::TypeDefinition::TypeDefinition):
(JSC::Wasm::TypeDefinition::allocatedRecursionGroupSize):
(JSC::Wasm::TypeDefinition::allocatedProjectionSize):
* Source/JavaScriptCore/wasm/WasmTypeDefinitionInlines.h:
(JSC::Wasm::TypeInformation::getFunctionSignature):
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):
* Source/JavaScriptCore/wasm/wasm.json:

Canonical link: https://commits.webkit.org/253491@main
